c++/src/kj/compat/tls.c++

```diff
@@ -312,78 +317,57 @@ private:
-    return sslCall([this,buffer,maxBytes]() { return SSL_read(ssl, buffer, maxBytes); })
-        .then([this,buffer,minBytes,maxBytes,alreadyDone](size_t n) -> kj::Promise<size_t> {
-      if (n >= minBytes || n == 0) {
-        return alreadyDone + n;
-      } else {
-        return tryReadInternal(reinterpret_cast<byte*>(buffer) + n,
-                               minBytes - n, maxBytes - n, alreadyDone + n);
-      }
-    });
-  }
-
-  Promise<void> writeInternal(kj::ArrayPtr<const byte> first,
-                              kj::ArrayPtr<const kj::ArrayPtr<const byte>> rest) {
+  Promise<void> writeInternal(kj::ArrayPtr<const kj::ArrayPtr<const byte>> slices) {
     KJ_REQUIRE(shutdownTask == kj::none, "already called shutdownWrite()");
 
-    // SSL_write() with a zero-sized input returns 0, but a 0 return is documented as indicating
-    // an error. So, we need to avoid zero-sized writes entirely.
-    while (first.size() == 0) {
-      if (rest.size() == 0) {
-        return kj::READY_NOW;
+    auto cork = writeBuffer.cork();
+    for (auto slice: slices) {
+      // SSL_write() with a zero-sized input returns 0, but a 0 return is documented as indicating
+      // an error. So, we need to avoid zero-sized writes entirely.
+      if (slice.size() == 0) continue;
+
+      while (slice.size() > 0) {
+        auto n = co_await sslCall([this,slice]() { return SSL_write(ssl, slice.begin(), slice.size()); });
+        if (n == 0) {
+          throw KJ_EXCEPTION(DISCONNECTED, "ssl connection ended during write");
+        }
+        slice = slice.slice(n, slice.size());
       }
-      first = rest.front();
-      rest = rest.slice(1, rest.size());
     }
-
-    return sslCall([this,first]() { return SSL_write(ssl, first.begin(), first.size()); })
-        .then([this,first,rest](size_t n) -> kj::Promise<void> {
-      if (n == 0) {
-        return KJ_EXCEPTION(DISCONNECTED, "ssl connection ended during write");
-      } else if (n < first.size()) {
-        return writeInternal(first.slice(n, first.size()), rest);
-      } else if (rest.size() > 0) {
-        return writeInternal(rest[0], rest.slice(1, rest.size()));
-      } else {
-        return kj::READY_NOW;
-      }
-    });
   }
 
   template <typename Func>
-  kj::Promise<size_t> sslCall(Func&& func) {
-    auto result = func();
+  kj::Promise<size_t> sslCall(Func func) {
+    while (true) {
+      auto result = func();
+      if (result > 0) {
+        co_return result;
+      }
 
-    if (result > 0) {
-      return result;
-    } else {
       int error = SSL_get_error(ssl, result);
       switch (error) {
         case SSL_ERROR_ZERO_RETURN:
-          return constPromise<size_t, 0>();
+          co_return 0;
         case SSL_ERROR_WANT_READ:
-          return readBuffer.whenReady().then(
-              [this,func=kj::mv(func)]() mutable { return sslCall(kj::fwd<Func>(func)); });
+          co_await readBuffer.whenReady();
+          break;
```

Comment:
It would be much clearer to call `sslCall` explicitly here, so much so that I think whatever slowdown this causes is worth it.

Reply:
but this is a recursive call, and we can't have recursive calls with coro.